File: src/python/dsa/myarrayqueue.py
from typing import Any, List, Optional, TypeVar

from typing_extensions import override

from .myqueue import MyQueue

T = TypeVar("T")


class MyArrayQueue(MyQueue[T]):
    """An array queue data structure (FIFO)."""

    DEFAULT_CAPACITY = 16
    MIN_ARR_LENGTH_THRESHOLD = 100

    def __init__(self, initialCapacity: Optional[int] = None) -> None:
        """Constructs an array queue instance with a specified initial capacity.

        Args:
            initialCapacity (int): initial capacity of this array queue. Defaults to 16.

        Raises:
            ValueError: if initialCapacity is negative
        """
        if not initialCapacity:
            initialCapacity = self.DEFAULT_CAPACITY
        elif initialCapacity < 0:
            raise ValueError("Negative capacity provided")
        self._array = [None] * initialCapacity
        self._array_size = 0

    def clear(self) -> None:
        pass

    def equals(self, obj: Any) -> bool:
        return False

    @override
    def __eq__(self, obj: Any) -> bool:
        """Overrides the `__eq__` method in the `object` class. Calls `equals`.

        Args:
            obj (Any): object to compare to this array queue

        Returns:
            bool: result from `equals`
        """
        return self.equals(obj)

    def isEmpty(self) -> bool:
        return True

    def __len__(self) -> int:
        """Returns the number of elements in this array queue.

        Returns:
            int: size of array queue
        """
        return self.size()

    def offer(self, element: T) -> None:
        pass

    def peek(self) -> Optional[T]:
        return None

    def poll(self) -> Optional[T]:
        return None

    def size(self) -> int:
        return 0

    def toArray(self) -> List[T]:
        return []

    def toString(self) -> str:
        return "[]"

    @override
    def __str__(self) -> str:
        """Overrides the `__str__` method in the `object` class. Calls `toString`.

        Returns:
            str: result from `toString`
        """
        return self.toString()

    def _checkCapacity(self) -> None:
        """Doubles or halves the size of the internal array depending on size. Used in offer and poll methods."""

File: src/typescript/dsa/myarrayqueue.ts
import MyQueue from "./myqueue";

/**
 * An array queue data structure (FIFO).
 */
export default class MyArrayQueue<T> implements MyQueue<T> {
    discriminator = "QUEUE";
    private array: T[];
    private array_size: number;
    private static readonly defaultCapacity = 16;
    private readonly minArrLenThreshold = 100;

    /**
     * Constructs an array queue instance with a specified initial capacity.
     * @param initialCapacity initial capacity of this array list. Defaults to
     * 16.
     */
    constructor(initialCapacity?: number) {
        if (initialCapacity === undefined) {
            initialCapacity = MyArrayQueue.defaultCapacity;
        } else if (initialCapacity < 0) {
            throw new Error("Negative capacity provided");
        }
        this.array = new Array<T>(initialCapacity);
        this.array_size = 0;
    }

    clear(): void {

    }

    equals(obj: object): boolean {
        return false;
    }

    isEmpty(): boolean {
        return true;
    }

    offer(element: T): void {

    }

    peek(): T {
        return null;
    }

    poll(): T {
        return null;
    }

    size(): number {
        return 0;
    }

    toArray(): T[] {
        return new Array<T>();
    }

    toString(): string {
        return "[]";
    }

    /**
     * Doubles or halves the size of the internal array depending on size.
     * Used in offer and poll methods.
     */
    private checkCapacity(): void {

    }
}
